systemservices/marketplace/src/contracts/utils.ts

```diff
@@ -42,22 +42,26 @@ interface CreateTransaction {
 const createTransactionTemplate = (
   chainID: number,
   web3: Web3,
-  defaultGas: number,
   defaultGasPrice: number
 ): CreateTransaction => async (
   contract: Contract,
   inputs: TaskInputs,
   data: string,
   shiftNonce?: number
-) => ({
-  chainID: chainID,
-  to: contract.options.address,
-  nonce: (await web3.eth.getTransactionCount(inputs.from)) + (shiftNonce || 0),
-  gas: inputs.gas || defaultGas,
-  gasPrice: inputs.gasPrice || defaultGasPrice,
-  value: '0',
-  data: data
-})
+) => {
+  const tx = {
```

Comment:
Not something blocking but this is quite messy, I think `estimateGas` might need the `nonce`, `gasPrice` and `gas` to run the simulation, if it doesn't need it it probably just ignore it so we could write something like that that feels easier to understand
```js
const tx = { chainId, to: ..., value: 0, data, nonce: ..., gasPrice: ..., gas: inputs.gas }
if (!tx.gas) {
  tx.gas = await web3.eth.estimateGas({ ...tx, from: inputs.from })
}
return tx
```

Reply:
I updated the code to your suggestion.